src/util/result.h

```diff
@@ -8,16 +8,135 @@
 #include <attributes.h>
 #include <util/translation.h>
 
+#include <memory>
+#include <optional>
+#include <tuple>
 #include <variant>
+#include <vector>
 
 namespace util {
+namespace detail {
+//! Empty string list
+const std::vector<bilingual_str> EMPTY_LIST{};
 
+//! Helper to move elements from one container to another.
+template <typename T>
+void MoveElements(T& src, T& dest)
+{
+    dest.insert(dest.end(), std::make_move_iterator(src.begin()), std::make_move_iterator(src.end()));
+    src.clear();
+}
+
+//! Error information only allocated if there are errors or warnings.
+template <typename F>
+struct ErrorInfo {
+    std::optional<std::conditional_t<std::is_same<F, void>::value, std::monostate, F>> failure;
+    std::vector<bilingual_str> errors;
+    std::vector<bilingual_str> warnings;
+};
+
+//! Result base class which is inherited by Result<T, F>.
+template <typename T, typename F>
+class ResultBase;
+
+//! Result base specialization for empty (T=void) value type. Holds error
+//! information and provides accessor methods.
+template <typename F>
+class ResultBase<void, F>
+{
+protected:
+    std::unique_ptr<ErrorInfo<F>> m_info;
+
+    template <typename... Args>
+    void InitFailure(Args&&... args)
+    {
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->failure.emplace(std::forward<Args>(args)...);
+    }
+    void InitValue() {}
+    template <typename O>
+    void MoveValue(O& other) {}
+    void DestroyValue() {}
+
+public:
+    void AddError(bilingual_str error)
+    {
+        if (error.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->errors.emplace_back(std::move(error));
+    }
+
+    void AddWarning(bilingual_str warning)
+    {
+        if (warning.empty()) return;
+        if (!m_info) m_info = std::make_unique<ErrorInfo<F>>();
+        m_info->warnings.emplace_back(std::move(warning));
+    }
+
+    //! Success check.
+    operator bool() const { return !m_info || !m_info->failure; }
```

Comment:
Would it be sufficient to use `explicit operator bool() const` instead here? That would avoid e.g. "(result + 3)" from compiling (and evaluating to 3 or 4).

Reply:
+1 to @sipa 's comment.